posts/inside-rust/2024-08-09-async-closures-call-for-testing.md

```diff
@@ -0,0 +1,171 @@
+    // Do things here...
+});
+```
+
+And on the callee side, write async fn trait bounds instead of writing "regular" fn trait bounds that return futures:
+
+```rust
+// Instead of writing:
+fn doesnt_exactly_take_an_async_closure<F, Fut>(callback: F)
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = String>
+{ todo!() }
+
+// Write this:
+fn takes_an_async_closure<F: async FnOnce() -> String>(callback: F) { todo!() }
+// Or this:
+fn takes_an_async_closure<F: AsyncFnOnce() -> String>(callback: F) { todo!() }
+```
+
+Or if you're emulating a higher-ranked async closure with boxing:
+
+```rust
+// Instead of writing:
+fn higher_ranked<F>(callback: F)
+where
+    F: Fn(&Arg) -> Pin<Box<dyn Future<Output = ()> + '_>>
+{ todo!() }
+
+// Write this:
+fn higher_ranked<F: async Fn(&Arg)> { todo!() }
+// Or this:
+fn higher_ranked<F: AsyncFn(&Arg)> { todo!() }
+```
+
+## Shortcomings interacting with the async ecosystem
+
+If you're going to try to rewrite your async projects, there are a few shortcomings to be aware of.
+
+### You can't directly name the output future
+
+When you name an async callable bound with the *old* style, before first-class async fn trait bounds, then as a side-effect of needing to use two type parameters, you can put additional bounds (e.g. `+ Send` or `+ 'static`) on the `Future` part of the bound, like:
+
+```rust
+fn async_callback<F, Fut>(callback: F)
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = String> + Send + 'static
+{ todo!() }
+```
+
+There isn't currently a way to put similar bounds on the future returned by calling an async closure, so if you need to constrain your callback futures like this, then you won't be able to use async closures just yet.
+
+We expect to support this in the medium/long term via a [return-type-notation syntax](https://rust-lang.github.io/rfcs/3668-async-closures.html#interaction-with-return-type-notation-naming-the-future-returned-by-calling).
+
+### Subtle differences in closure signature inference
+
+Passing an async closure to a generic `impl Fn(A, B) -> C` bound may not always eagerly infer the closure's arguments to `A` and `B`, leading to strange type errors on occasion. For an example of this, see [`rust-lang/rust#127781`](https://github.com/rust-lang/rust/issues/127781).
+
+We expect to improve async closure signature inference as we move forward.
+
+### Async closures can't be coerced to `fn()` pointers
+
+Some libraries take their callbacks as function *pointers* (`fn()`) rather than generics. Async closures don't currently implement the same coercion from closure to `fn() -> ...`. Some libraries may mitigate this problem by adapting their API to take generic `impl Fn()` instead of `fn()` pointers as an argument.
+
+We don't expect to implement this coercion unless there's a particularly good reason to support it, since this can usually be handled manually by the caller by using an inner function item, or by using an `Fn` bound instead: for example:
```

Comment:
```suggestion
We don't expect to implement this coercion unless there's a particularly good reason to support it, since this can usually be handled manually by the caller by using an inner function item, or by using an `Fn` bound instead, for example:
```